### ISSUE TITLE
flume 0.8.4 panics on parallel recv_async()

### ISSUE THREAD

**Issue by hatoo** [zesterer/flume#42]
Hi, Thanks for the nice project!
During creating an application with `flume` I found a panic in debug build.

full codes are https://github.com/hatoo/flume-panic

# My Environment
- WSL2 Ubuntu20.04

```rust
use futures::stream::FuturesUnordered;
use futures::StreamExt;

#[tokio::main]
async fn main() {
    let (tx, rx) = flume::unbounded();
    tokio::spawn(async move {
        let n_sends = 100000;
        for _ in 0..n_sends {
            tx.send_async(()).await.unwrap();
        }
        println!("send end");
    });

    let mut futures_unordered = (0..250)
        .map(|_| async {
            while let Ok(()) = rx.recv_async().await
            /* rx.recv() is OK */
            {}
        })
        .collect::<FuturesUnordered<_>>();

    while futures_unordered.next().await.is_some() {}
    println!("recv end");
}
```

The program sends many values to the channel and receives via parallel workers with `recv_async()` and panics in my environment. The log for `RUST_BACKTRACE=full` is pasted in https://github.com/hatoo/flume-panic.
Also, I've found that it works correctly if change `recv_async().await` to `recv()`.

So I'm guessing there is an issue in recv_async().

Thanks.

**Comment by Restioson** [comment]
I encountered this too and it's fixed on #40

**Comment by Restioson** [comment]
I think the fix was just to remove the assertion since after some investigation it doesn't seem to be a valid assertion with the way we do async anyways.

**Comment by hatoo** [comment]
I tested #40 on https://github.com/hatoo/flume-panic/tree/pr-40

```toml
flume = { git = "https://github.com/Restioson/flume.git", branch = "async-edits" }
```

And got another panic and prints `recv end` on debug build.
And freezes and prints `send end` on release build.

I've found that it's OK if change `send_async` to `send`.

Is `send_async` has same kind issue?

**Comment by Restioson** [comment]
Apologies, it appears like I had not pushed the change at the time!

**Comment by Restioson** [comment]
Okay, I believe I found the issue. I am going to push a fix soon.

**Comment by hatoo** [comment]
I've confirmed it works on 6f318d5.
Thank you!

**Comment by Restioson** [comment]
Thanks! If it's alright with you, could I integrate your example as above into the test suite?

**Comment by hatoo** [comment]
Yes of course!

**Comment by zesterer** [comment]
I've just published `0.9.0` which includes these fixes.

**Comment by hatoo** [comment]
Thanks. It worked!
